flow/ProtocolVersion.h

```diff
@@ -0,0 +1,123 @@
+// the `ProtocolVersion`-class (through the template parameter) and it has
+// to be names `P` (otherwise the VERSION_FEATURE macro won't work anymore).
+// These classes are used to determine if a feature was available at a certain
+// protocol version. It does determine that through static polymorphism.
+template<class P>
+struct v5_5 {
+	static constexpr uint64_t protocolVersion = 0x0FDB00A551000000LL;
+
+	VERSION_FEATURE(MultiVersionClient)
+};
+
+template<class P>
+struct v5_6 : v5_5<P> {
+	static constexpr uint64_t protocolVersion = 0x0FDB00A560010001LL;
+
+	VERSION_FEATURE(TagLocality)
+};
+
+template <class P>
+struct v6_0 : v5_6<P> {
+	static constexpr uint64_t protocolVersion = 0x0FDB00B060000001LL;
+
+	VERSION_FEATURE(Fearless)
+};
+
+template <class P>
+struct v6_1 : v6_0<P> {};
+
+// This typedef needs to be updated whenever a new version is added
+template <class P>
+using latest_version = v6_1<P>;
+
+} // namespace fdb_versions
+
+// ProtocolVersion wraps a uint64_t to make it type safe. It has to inherit
+// from the newest version object and it will know about the current version.
+// The default constuctor will initialize the version to 0 (which is an invalid
+// version). ProtocolVersion objects should never be compared to version numbers
+// directly. Instead one should always use the type-safe version types from which
+// this class inherits all.
+class ProtocolVersion : public fdb_versions::latest_version<ProtocolVersion> {
+	uint64_t _version;
+public: // constants
+	static constexpr uint64_t versionFlagMask = 0x0FFFFFFFFFFFFFFFLL;
+	static constexpr uint64_t objectSerializerFlag = 0x1000000000000000LL;
+	static constexpr uint64_t compatibleProtocolVersionMask = 0xffffffffffff0000LL;
+	static constexpr uint64_t minValidProtocolVersion = 0x0FDB00A200060001LL;
+
+public:
+	constexpr explicit ProtocolVersion(uint64_t version) : _version(version) {}
+	constexpr ProtocolVersion() : _version(0) {}
+
+	constexpr bool isCompatible(ProtocolVersion other) const {
+		return (other.version() & compatibleProtocolVersionMask) == (version() & compatibleProtocolVersionMask);
+	}
+	constexpr bool isValid() const { return version() >= minValidProtocolVersion; }
+
+	constexpr uint64_t version() const { return _version & versionFlagMask; }
+	constexpr uint64_t versionWithFlags() const { return _version; }
+
+	constexpr bool hasObjectSerializerFlag() const { return (_version & objectSerializerFlag) > 0; }
+	constexpr void addObjectSerializerFlag() { _version = _version | objectSerializerFlag; }
+	constexpr void removeObjectSerializerFlag() {
+		_version = hasObjectSerializerFlag() ? _version ^ objectSerializerFlag : _version;
+	}
+	constexpr void removeAllFlags() { _version = version(); }
+
+	constexpr operator bool() const { return _version != 0; }
+	// comparison operators
+	constexpr bool operator==(const ProtocolVersion other) const { return version() == other.version(); }
```

Comment:
This is more a thought than a request, but would it be helpful to have somewhere an explanation that flags don't affect comparisons between protocol versions and why?

Reply:
done